repository_service_tuf_worker/services/keyvault/local.py

```diff
@@ -76,33 +82,21 @@ def settings(cls) -> List[ServiceSettings]:
             ),
         ]
 
-    def get(self, rolename: str) -> Dict[str, Any]:
-        """Get the Key from local KeyVault by role name."""
-        keys_sslib_format: List[Dict[str, Any]] = []
+    def get_signer(self) -> SSlibSigner:
+        """Return a signer using the online key."""
         try:
-            keys: Dict[str, Any] = self.keyvault.store[rolename]
-            for key in keys:
-                keys_sslib_format.append(
-                    decrypt_key(key["key"], key["password"])
-                )
-        except (BoxKeyError, KeyError):
-            raise KeyVaultError(f"{rolename} key(s) not found.")
-
-        return keys_sslib_format
-
-    def put(self, rolename: str, keys: List[Dict[str, Any]]) -> None:
-        """Save the Key in the local KeyVault."""
-        key_vault_data: list = []
-        for key in keys:
-            ed25519_key = encrypt_key(key.get("key"), key.get("password"))
-            key_vault_data.append(
-                {
-                    "key": ed25519_key,
-                    "filename": key["filename"].split("/")[-1],
-                    "password": key["password"],
-                }
+            key_info: Dict[str, Any] = import_privatekey_from_file(
+                self._online_key_name,
+                self._online_key_type,
+                self._online_key_password,
             )
-
-        self.keyvault.store[rolename.upper()] = key_vault_data
-        data = self.keyvault.as_dict(env=self.keyvault.current_env)
-        loaders.write(self._secrets_file, DynaBox(data).to_dict())
+            return SSlibSigner(key_info)
```

Comment:
I still think your RSTUF KeyVault abstraction over the secureystemslib signer interface should take a public key and use from_priv_key_uri. Then you'd be able to reuse the same code for any Signer implementation.

I understand that you wanted to avoid passing in a public key,  which is possible for SSlibSigner, but will be harder for something like a cloud signer (e.g. GCPSigner), where the signer doesn't have direct access to the key material.

Plus it should not be a problem to get the public key for the signer. You can even grab it from root or top-level targets. :)


cc @jku

Reply:
I deleted by mistake my comment

Each RSTUF KeyVault abstraction (service) will define its [required parameters.](https://repository-service-tuf.readthedocs.io/en/latest/guide/repository-service-tuf-worker/Docker_README.html#required-rstuf-keyvault-backend) (required or optional)

It could be difficult to predict all the SaaS/Vault required parameters.

I will simulate different examples here:

* RSTUF_KEYVAULT_BACKEND: LocalKeyVault
  - RSTUF_LOCAL_KEYVAULT_PATH: /volume/path/
  - RSTUF_LOCAL_KEYVAULT_NAME: rstuf-online.key
  - RSTUF_LOCAL_KEYVAULT_TYPE: rsa
  - RSTUF_LOCAL_KEYVAULT_PASSWORD: /run/secrets/RSTUF_ONLINE_KEY

* RSTUF_KEYVAULT_BACKEND: GCPKM
  - RSTUF_GCPKM_KEYVAULT_PUBLIC_KEY: "-----BEGIN PUBLIC KEY----- ..."
  - RSTUF_GCPKM_KEYVAULT_PUBLIC_URI: "gcp-kms://projects/*/locations/*/keyRings/*/cryptoKeys/*"
 
* RSTUF_KEYVAULT_BACKEND: AWSKMS
  - RSTUF_AWSKMS_KEYVAULT_KEY_ID: "arn:aws:kms:<region>:<account-id>:key/<key-id>"

Reply:
I'm definitely not telling you what makes sense, I'm not familiar with this code, but since lukas pinged... This is how I would handle your examples:
* GCP: Use GCPSigner
* AWS: Implement AWSSigner in securesystemslib (with a private key uri that contains the AWS keyid), then use it
* LocalKeyVault: Implement own signer (maybe in application code?), with a private_key_uri containing all of the constant arguments you mentioned. If the arguments are not constant (or are secrets) you read them from a location that is constant and can be embedded in private_key_uri -- this is what the other signers do as well

Reply:
My main point here is about the public key. All current signer implementations need it, and RSTUF has it (via root metadata). So why not use that?

I don't think it's necessary to find a custom solution for every signer, such as passing it via a config setting (`RSTUF_GCPKM_KEYVAULT_PUBLIC_KEY`), or working around it by inferring it from the private key material (`SSlibSigner`).

Reply:
I see the benefit of passing the public key and using `from_priv_key_uri` directly.
For us, the biggest benefit is that we can easily have #202 implemented by redefining `KEY_FOR_TYPE_AND_SCHEME` however we want and using `Key. from_dict()`